File: Statistics/lesson1/sources/Lesson1VBNET/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace Lesson1CS
    Public Partial Class Lesson1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private running As Boolean = False
        Private populated As Boolean = False
        Private timer As Timer = New Timer()
        Private balls As List(Of Ball) = New List(Of Ball)()
        Private ball1 As Ball = New Ball(20, 20, 0, 0, 24, 24)
        Private ball2 As Ball = New Ball(30, 30, 10, 10, 20, 20)
        Private ball3 As Ball = New Ball(40, 40, 20, 20, 16, 16)
        Private ball4 As Ball = New Ball(50, 50, 30, 30, 12, 12)
        Private ball5 As Ball = New Ball(60, 60, 40, 40, 8, 8)

        Private Sub populateBalls()
            pictureBox1.BackColor = Color.White
            balls.Add(ball1)
            balls.Add(ball2)
            balls.Add(ball3)
            balls.Add(ball4)
            balls.Add(ball5)
        End Sub

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs)
            richTextBox1.Text = "Hello World! Click the button below to animate the balls, click again to stop them."
            richTextBox1.SelectionStart = 0
            richTextBox1.SelectionLength = richTextBox1.Text.Length
            richTextBox1.SelectionFont = New Font("Maiandra GD", 13)
        End Sub

        Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs)
            If Not populated Then
                populateBalls()
                populated = True
            End If

            If Not running Then
                running = True
                AddHandler timer.Tick, New EventHandler(AddressOf timer_Tick)
                timer.Interval = 10
                timer.Start()
            Else
                timer.[Stop]()
                running = False
            End If
        End Sub

        Private Sub timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            For i As Integer = 0 To balls.Count - 1
                updateBall(sender, e, balls(i))
            Next

            pictureBox1.Paint += New PaintEventHandler(AddressOf DrawCircles)
        End Sub

        Private Sub updateBall(ByVal sender As Object, ByVal e As EventArgs, ByVal b As Ball)
            Dim width As Integer = b.GetWidth()
            Dim height As Integer = b.GetHeight()
            Dim x As Integer = b.GetX()
            Dim y As Integer = b.GetY()
            Dim stepx As Integer = b.GetStepX()
            Dim stepy As Integer = b.GetStepY()
            x += stepx

            If x + width > pictureBox1.Width OrElse x < 0 Then
                stepx = -stepx
            End If

            y += stepy

            If y + height > pictureBox1.Height OrElse y < 0 Then
                stepy = -stepy
            End If

            b.SetWidth(width)
            b.SetHeight(height)
            b.SetX(x)
            b.SetY(y)
            b.SetStepX(stepx)
            b.SetStepY(stepy)
            pictureBox1.Refresh()
        End Sub

        Private Sub DrawCircles(ByVal sender As Object, ByVal e As PaintEventArgs)
            e.Graphics.Clear(BackColor)
            e.Graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.AntiAlias

            For i As Integer = 0 To balls.Count - 1
                Dim width As Integer = balls(i).GetWidth()
                Dim height As Integer = balls(i).GetHeight()
                Dim x As Integer = balls(i).GetX()
                Dim y As Integer = balls(i).GetY()
                e.Graphics.FillEllipse(Brushes.Yellow, x, y, width, height)
                e.Graphics.DrawEllipse(Pens.Black, x, y, width, height)
            Next
        End Sub
    End Class
End Namespace

File: Statistics/lesson1/sources/Lesson1CS/Lesson1CS/Form1.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace Lesson1CS
{
    public partial class Lesson1 : Form
    {
        public Lesson1()
        {
            InitializeComponent();
        }

        private bool running = false;
        private bool populated = false;
        private Timer timer = new Timer();
        List<Ball> balls = new List<Ball>();
        Ball ball1 = new Ball(20, 20, 0, 0, 24, 24);
        Ball ball2 = new Ball(30, 30, 10, 10, 20, 20);
        Ball ball3 = new Ball(40, 40, 20, 20, 16, 16);
        Ball ball4 = new Ball(50, 50, 30, 30, 12, 12);
        Ball ball5 = new Ball(60, 60, 40, 40, 8, 8);

        private void populateBalls()
        {
            pictureBox1.BackColor = Color.White;
            balls.Add(ball1);
            balls.Add(ball2);
            balls.Add(ball3);
            balls.Add(ball4);
            balls.Add(ball5);
        }

        // This function handles the add text functionality
        private void button1_Click(object sender, EventArgs e)
        {
            richTextBox1.Text = "Hello World! Click the button below to animate the balls, click again to stop them.";
            richTextBox1.SelectionStart = 0;
            richTextBox1.SelectionLength = richTextBox1.Text.Length;
            richTextBox1.SelectionFont = new Font("Maiandra GD", 13);
        }

        // This function handles the balls movement functionality
        private void button2_Click(object sender, EventArgs e)
        {
            if (!populated)
            {
                populateBalls();
                populated = true;
            }
            if (!running)
            {
                running = true;
                timer.Tick += new EventHandler(timer_Tick);
                // Sets the timer interval to 10 milliseconds.
                timer.Interval = 10;
                timer.Start();
            }
            else
            {
                timer.Stop();
                running = false;
            }
        }

        private void timer_Tick(object sender, EventArgs e)
        {
            for (int i = 0; i < balls.Count; i++)
            {
                updateBall(sender, e, balls[i]);
            }
            // Connect the Paint event of the PictureBox to the event handler method.
            pictureBox1.Paint += new PaintEventHandler(DrawCircles);
        }

        // this function updates the position of the balls.
        private void updateBall(object sender, EventArgs e, Ball b)
        {
            int width = b.GetWidth();
            int height = b.GetHeight();
            int x = b.GetX();
            int y = b.GetY();
            int stepx = b.GetStepX();
            int stepy = b.GetStepY();

            x += stepx;

            if (x + width > pictureBox1.Width || x < 0)
            {
                stepx = -stepx;
            }

            y += stepy;

            if (y + height > pictureBox1.Height || y < 0)
            {
                stepy = -stepy;
            }

            b.SetWidth(width);
            b.SetHeight(height);
            b.SetX(x);
            b.SetY(y);
            b.SetStepX(stepx);
            b.SetStepY(stepy);

            pictureBox1.Refresh();
        }

        private void DrawCircles(object sender, PaintEventArgs e)
        {
            e.Graphics.Clear(BackColor);
            e.Graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.AntiAlias;
            for (int i = 0; i < balls.Count; i++)
            {
                int width = balls[i].GetWidth();
                int height = balls[i].GetHeight();
                int x = balls[i].GetX();
                int y = balls[i].GetY();
                e.Graphics.FillEllipse(Brushes.Yellow,
                x, y,
                width, height);
                e.Graphics.DrawEllipse(Pens.Black,
                    x, y,
                    width, height);
            }
        }
    }
}